Add removal reason for BlockEntity#setRemoved

sciwhiz12 opened the pull request:
This PR closes #8302 by implementing an overload for `BlockEntity#setRemoved` with a parameter for the **removal reason** for the `BlockEntity`, modeled after `Entity.RemovalReason`. The enum `IForgeBlockEntity.RemovalReason` provides the four removal reasons that may be passed into the new overload, with documentation as to what they represent and where they are called.

The original `setRemoved()` is kept for compatibility reasons, and redirects to `setRemoved(RemovalReason)` with the catch-all `REMOVED` reason. The new overload is specified in the extension interface for `BlockEntity` for the javadoc.

Furthermore, the `IForgeBlockEntity#onChunkUnloaded` method's call path was adjusted to account for the new removal reason, ensuring that the method is called only upon the chunk being actually unloaded, rather than both due to chunk unload and due to chunk packet synchronization.

sciwhiz12 committed:
Add removal reason for BlockEntity#setRemoved

Implements and closes #8302

sciwhiz12 committed:
Call BlockEntity#onChunkUnloaded only on unloaded reason

sciwhiz12 committed:
Cleanup formatting

sciwhiz12 committed:
Split REMOVED reason into three separate reasons

sciwhiz12 committed:
Terminally deprecate #onChunkUnloaded

sciwhiz12 committed:
Commit forgotten patches

sciwhiz12 committed:
Add isChunk() to the removal reasons

This allows differentiating between reasons that target the specific
block entity versus reasons that target the chunk the block entity
resides in.

sciwhiz12 committed:
Reorder CLEARED_CHUNK to be near other chunk-related reasons

Technici4n commented on src/main/java/net/minecraftforge/common/extensions/IForgeBlockEntity.java:
```diff
@@ -189,4 +196,69 @@ default void requestModelDataUpdate()
+    /**
+     * Reasons for the removal of a {@link BlockEntity} from the level.
+     */
+    enum RemovalReason
```
I'm not sure all these cases, and the `boolean chunk` parameter, are useful. Wouldn't REMOVED and UNLOADED be enough, possibly UNKNOWN too?

sciwhiz12 replied:
The various removal reasons were added in an effort to provide as much information to the modder as reasonably possible. We certainly could cut down the amount of values in this enum (though for two values, it'd be better then to use a `boolean`), but I felt at the time that providing the info would be of some potential use and wasn't too hard to implement at least.

Technici4n replied:
More info means more cases to handle for the modder though. I think that most use cases would be covered by a simple boolean `wasUnloaded`, probably with a default of false.

pupnewfster replied:
I am inclined to agree that at least adding a second boolean to `RemovalReason` for removed (and not coming back) vs just unloaded would make sense and be useful for devs (`REMOVED` and `REPLACED_IN_WORLD`, not sure about `UNKNOWN` and I haven't looked into how `CLEARED_CHUNK` is used)

sciwhiz12 replied:
@pupnewfster: That sounds like a reasonable idea to me, so modders can rely on the boolean instead of hardcoding the individual reasons (and possibly needing to add any further new reasons should they come up). This way, we retain the information provided by the individual `RemovalReason`s (which was my original reason for the enum) and provide a simple boolean check for modders to use (as @Technici4n notes).

sciwhiz12 committed:
Add missing override annotation

sciwhiz12 committed:
Add removedPermanently boolean to RemovalReason

sciwhiz12 committed:
Remove impl of onChunkUnloaded

The capabilities are already invalidated in setRemoved(RemovalReason),
so this needlessly doubles the calls to invalidation listeners when a
chunk is unloaded.

sciwhiz12 committed:
Remove UNKNOWN removal reason with REMOVED

As the UNKNOWN reason serves little purpose other than currently a
placeholder value when calling the original setRemoved, it was decided
that it should be removed and be replaced with the REMOVED reason.

This means that callers of the original setRemoved() now signal that
the block entity is removed permanently. However, it is noted that this
would only be for existing mods; as the original setRemoved() is
deprecated, mods shouldn't be using it and should prefer the one with
the explicit removal reason.

sciwhiz12 committed:
Override setRemoved(RemovalReason) in BeaconBlockEntity

The thought of patching this method further to ensure that the beacon
deactivation sound only fires for the REMOVED reason. However, this was
deferred for a future PR with its own discussion, as that would be
changing vanilla behavior and out of scope for this PR.

sciwhiz12 committed:
Replace uses of setRemoved() in PistonMovingBlockEntity

sciwhiz12 committed:
Remove manual setRemoved in BE erroring removal code

The removal is already signalled in LevelChunk.removeBlockEntity, so
this needlessly duplicates it.

LexManos commented:
Quick look over and discussion on discord, I think this needs more work and looking into the usecases.
I think {I don't know because I haven't seen enough actual uses} that this would fall more under the modders using the wrong method. It may be more appropriate to make these 'cleanup' things invoked from Block.onRemove. Which is the thing at the Block level that is responsible for cleaning up Block Entities when the blockstate changes in the chunk.

sciwhiz12 committed:
Merge latest 1.18.x
